pkg/gadgets/trace/tcpdrop/tracer/tracer.go

```diff
@@ -0,0 +1,208 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tracer
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	"golang.org/x/sys/unix"
+
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/tcpdrop/types"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/tcpbits"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target $TARGET -cc clang -no-global-types -type event tcpdrop ./bpf/tcpdrop.bpf.c -- -I./bpf/ -I../../../../${TARGET}
+
+type Tracer struct {
+	dropReasons map[int]string
+
+	eventCallback func(*types.Event)
+
+	objs         tcpdropObjects
+	kfreeSkbLink link.Link
+	reader       *perf.Reader
+}
+
+func (t *Tracer) close() {
+	t.kfreeSkbLink = gadgets.CloseLink(t.kfreeSkbLink)
+
+	if t.reader != nil {
+		t.reader.Close()
+	}
+
+	t.objs.Close()
+}
+
+func (t *Tracer) loadDropReasons() error {
+	btfSpec, err := btf.LoadKernelSpec()
+	if err != nil {
+		return fmt.Errorf("loading kernel spec: %w", err)
+	}
+
+	t.dropReasons = make(map[int]string)
+	enum := &btf.Enum{}
+	err = btfSpec.TypeByName("skb_drop_reason", &enum)
+	if err != nil {
+		return fmt.Errorf("looking up skb_drop_reason enum: %w", err)
+	}
+	for _, v := range enum.Values {
+		str := v.Name
+		str = strings.TrimPrefix(str, "SKB_DROP_REASON_")
+		str = strings.TrimPrefix(str, "SKB_")
+
+		t.dropReasons[int(v.Value)] = str
+	}
+
+	return nil
+}
+
+func (t *Tracer) install() error {
+	err := t.loadDropReasons()
+	if err != nil {
+		return err
+	}
+
+	spec, err := loadTcpdrop()
+	if err != nil {
+		return fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	gadgets.FixBpfKtimeGetBootNs(spec.Programs)
+
+	opts := ebpf.CollectionOptions{}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		return fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	t.kfreeSkbLink, err = link.Tracepoint("skb", "kfree_skb", t.objs.IgTcpdrop, nil)
+	if err != nil {
+		return fmt.Errorf("opening tracepoint kfree_skb: %w", err)
+	}
+
+	reader, err := perf.NewReader(t.objs.tcpdropMaps.Events, gadgets.PerfBufferPages*os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("creating perf ring buffer: %w", err)
+	}
+	t.reader = reader
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	for {
+		record, err := t.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				// nothing to do, we're done
+				return
+			}
+
+			msg := fmt.Sprintf("reading perf ring buffer: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
```

Comment:
Instead of using events for log messages, you can now also use the logging facility from the `gadgetCtx` now. The advantage of using that, is that messages can already be filtered on the server using the log level.

Keep in mind though, that this is not compatible with the CR path (which isn't used here), yet (gadget-collection doesn't support it, because it doesn't build the gadget context).

Reply:
Since other gadgets don't do this yet, I suggest to do the transition in a separate PR for all gadgets.